### DISCUSSION TITLE
Question about Detail Communication Cost and Memory Cost of Multi-Dimensional Parallelism

### DISCUSSION THREAD

**Discussion by ConnollyLeon** [hpcaitech/ColossalAI#185] [Community | Q&A] [answered]
Dear developers,

I have read the papers of 2D, 2.5D and 3D parallelism, but still have some questions about the communication cost and memory cost of Colossal-AI. 

Here is my question list:

In  3D parallelism,  I notice that Colossal-AI moves `(MN +NK +MK)/p^3`  across `p` devices in a matrix-matrix multiplication using All-gather or Reduce-scatter. And the final communication bandwidth cost is `O(1/p^2)`. 

- Q1: Is this communication implemented by RING, whose communication bandwidth cost is specifically `(MN +NK +MK)*(p-1)/p^3`?
- Q2: All-gathering matrix `A` and `B` might temporally create sub-matrix with shape of `(M/p, N/p)` and `(N/p, K/p)`. Although it can be dropped after reduce-scattering matrix `C`, I still think the peak memory cost of matrix multiplication turns out to be `O(1/p^2) `instead of `O(1/p^3)` in the paper. I wonder if the `O(1/p^3)` is described for the whole process of an iteration. 
- Q3: I am still a green hand in communication field, and I notice that in the paper, you say the communication latency of 3D parallelism is `O(log(p))`. Could  you please tell me how this latency is calculated?
- Q4: How do you organize the communication well since the bandwidth between intra-node devices and inter-node devices are different? Although the communication volume is highly reduced, would this be a bottleneck in 2D or 3D parallelism?

Waiting for your responses, please help us~~

**Comment by FrankLeeeee** [comment]
Hi @ConnollyLeon , thank you for your question. @kurisusnowdeng is in charge of 3D parallelism. He will try to answer your question. :)

**Comment by kurisusnowdeng** [comment, marked as answer]
Hi, @ConnollyLeon . Thank you for your interest. I give my anwsers according to my understanding to your questions one by one.

- A1: If using NCCL as the communication backend, the RING algorithm is supposed to be used, and the bandwidth cost you gave is correct.
- A2: Indeed during both communication and multiplication the memory cost is O(1/p^2). However, such memory can be reused by other layers. Compared to the memory that is kept for backward (usually tens of transformer layers), it seems not so significant in terms of peak memory usage. That is why we also partition the activations by p^3.
- A3: If using a tree topology for parallel communication, the latency can be O(log(p)). You can refer to [these slides](https://developer.download.nvidia.cn/assets/cuda/files/reduction.pdf) where we first found the conclusion. However, with the RING topology, the latency should be O(p-1). Thank you for correction.
- A4: Communication across different nodes can be costly. We now suggest to use tensor parallelism within a single server, such as 8x GPUs for 2x2x2 3D parallelism. We are also making our efforts to reduce the communication cost in order to improve the throughput.

**Reply by ConnollyLeon** [reply]
Thanks for your answers, and most of them matches my guesses, but  I still have a small question about A4:

- You suggest to use TP within a single server, but in the 3D tensor parallelism paper, you give an experiment result of 64 GPUs (16 servers, each have 4 GPU), which have already outperformed 1D  and 2D. It seems like that it already has a good improvement in multi-node environment. But why did you suggest to use TP within a single server instead?  Could you please tell us whether there is any performance deterioration (Compared to 1D or 2D) in deploying 3D TP on multi-nodes? 

A little suggestion: It would be better to provide some quantized profiles of communication time and volume along different axes (which may have different bandwidth), to help community understand the advantages of Colossal-AI :).

**Reply by kurisusnowdeng** [reply]
Since the model size is increasing, we may have to use multiple servers to accommodate a huge transformer model, e.g. GPT-3. So, it is important to present the scalability of 3D parallelism even across servers. I mean we suggest not to use too many servers for TP if unnecessary. BTW, thank you so much for the suggestion.

**Reply by ConnollyLeon** [reply]
Thanks for your replies.  I try to  analyze the memory of 3D-TP  formally, could you please check whether I am right?

Suppose  we have `l` layers (or operators that has weights), and  `p=q*q*q` devices, and the memory of the model is `M`.  The total memory cost would be `(1 / q^2 * 1 / l) M + (1 / q^3 * (l - 1) / l) M = (q + l - 1) /( l * q^3)M`, which means that there is only one layer that requires to hold `(1 / q^2)` part during running.  When the `l` is much larger than `q`, the memory cost become s`O(1 / q^3)` according to L'Hôpital's rule. When the `l` equals to `1`, the memory cost becomes `O(1 / q^2)`.  So the interval of the memory cost of 3D-TP is` (1 / q^3 M, 1 / q^2 M]`. When given specific model with `l` layers, and `p` devices, the memory cost of each device can be easily calculated and is in this interval.

**Comment by ConnollyLeon** [comment]
@kurisusnowdeng 还有一个小问题，why not use All-gather instead  of Broadcast in 2D model parallelism?  According to my analysis, the total communication time (bandwidth cost specifically) of using All-gather would be smaller.

**Reply by FrankLeeeee** [reply]
> @kurisusnowdeng 还有一个小问题，why not use All-gather instead of Broadcast in 2D model parallelism? According to my analysis, the total communication time (bandwidth cost specifically) of using All-gather would be smaller.

We have tested using all-gather and broadcast. All-gather leads to higher memory cost. In terms of speed, they are quite close if using async communication.

**Reply by ConnollyLeon** [reply]
Thanks for your reply. I have read the implementation here. 

https://github.com/hpcaitech/ColossalAI/blob/f5ca88ec972f4c3c76e8aadaacce7a078d1f7116/colossalai/nn/layer/parallel_2d/_operation.py#L390 

It seems like that you replace `Broadcast of A` for `Reduce of C`, and uses a local buffer `B_list` and `C_list` to support the overlap of communication and computation. Using buffers can help reuse memory when compared to All-gather, I think the memory advantage comes from here.  But  why do you make this replacement? Would it be better to analyze the size of `A` and `C`  first to determine a `Broadcast of A` or a `Reduce of C` (pick the one with smaller size for  less communication amount)?

**Reply by FrankLeeeee** [reply]
This is a matmul of AB^T. According to [SUMMA Paper](http://www.netlib.org/lapack/lawnspdf/lawn96.pdf), reduction of C is needed.

<img width="315" alt="Screenshot 2022-02-18 at 10 40 36 AM" src="https://user-images.githubusercontent.com/31818963/154607457-2190e7bf-c750-4d99-afc5-5e07b03a0653.png">

**Reply by ConnollyLeon** [reply]
Thanks~, It  was  my fault not  reading  it carefully. Now I  completely understand. 😄

**Reply by FrankLeeeee** [reply]
Welcome! :)
